docs: update with more patterns and decision based (#90203)
vercel/next.js: docs/01-app/02-guides/activity-cache-components.mdx

When [Cache Components](/docs/app/getting-started/cache-components) is enabled, routes don't unmount when you navigate away from them. Instead, Next.js uses React's [`<Activity>`](https://react.dev/reference/react/Activity) component to hide them - setting them to `mode="hidden"` rather than removing them from the tree.

> This preserves component state and DOM state, which is great for back/forward navigation. But it also means you need to decide which state should persist and which should reset - patterns that relied on unmounting to clear state now need explicit choices.

Reviewer: > now need explicit choices
This is confusing

If you want to understand how Activity works and the patterns it enables, see the [Activity guide](/docs/app/guides/activity).

## Choosing what to preserve

Activity preserves all component state by default. For each piece of state, you decide whether that's the right behavior for your UI. The patterns below show common scenarios and how to handle both sides.

### Expandable UI (dropdowns, accordions, panels)

When a user navigates away and returns, Activity preserves the open/closed state of expandable elements.

**When to keep it:** A sidebar with expanded sections, a FAQ accordion, or a filters panel - the user set up their view intentionally, and restoring it avoids re-doing that work.

> **When to reset it:** A dropdown menu or popover triggered by a button click - these are transient interactions, not persistent view state. Returning to a page with a dropdown already open feels wrong.

Reviewer: ```suggestion
**When to reset it:** A dropdown menu or popover triggered by a button click - these are transient interactions, not persistent view state. Returning to a page with a dropdown already open is not user friendly.
```

}
```

If the user navigated away while the dialog was open, returning and opening the dialog won't trigger the focus Effect because `isDialogOpen` was already `true`.

> To fix this, derive the dialog state from something outside the preserved component state - like a search param:

Reviewer: ```suggestion
To fix this, derive the dialog state from something outside the preserved component state like a search param:
```
In general, avoid using "-" unless it makes grammatical sense i.e giving more details about what comes before as an interjection

This resets the form whenever the user navigates away. See [Resetting state](/docs/app/guides/activity#resetting-state) in the Activity guide for more patterns, including key-based resets for React state.

### Action state (`useActionState`)

Activity preserves [`useActionState`](https://react.dev/reference/react/useActionState) results - success messages, error messages, and any other state returned by the action.

> **When to keep it:** A ticket redemption form showing "Ticket redeemed successfully", or a settings form showing "Changes saved". Seeing the result of a previous action when returning to the page is useful confirmation - the user can see what happened.

Reviewer: ```suggestion
**When to keep it:** A ticket redemption form showing "Ticket redeemed successfully", or a settings form showing "Changes saved". Seeing the result of a previous action when returning to the page is useful confirmation so the user can see what happened.
```

  // Dispatch RESET when Activity hides this component
  useLayoutEffect(() => {
    return () => {
      if (shouldReset.current) {
        shouldReset.current = false
        startTransition(() => {
          dispatch({ type: 'RESET' })
        })
      }
    }
  }, [dispatch])

  return (
    <form action={(formData) => dispatch({ type: 'SUBMIT', data: formData })}>
      <textarea name="comment" />
      <button type="submit" disabled={isPending}>
        {isPending ? 'Posting...' : 'Post Comment'}
      </button>
      {state.success && <p>Comment posted!</p>}
      {state.error && <p>{state.error}</p>}
    </form>
  )
}
```

Here's what happens step by step:

1. The user submits the form. The reducer receives a `SUBMIT` action with the `FormData`, calls `saveComment`, and returns `{ success: true }`. It also sets `shouldReset.current = true` to mark that a reset is needed.
2. The user navigates away. Activity hides the component and runs the `useLayoutEffect` cleanup. Because `shouldReset.current` is `true`, it dispatches a `RESET` action.
3. The reducer receives `RESET` and returns the initial state (`{ success: false, error: null }`). The stale success message is cleared.

> 4. If the user navigates back, the form is ready for a new submission. If they never submitted (step 1 didn't happen), `shouldReset.current` is still `false`, so no `RESET` is dispatched - the form stays as-is.

Reviewer: ```suggestion
4. If the user navigates back, the form is ready for a new submission. If they never submitted (step 1 didn't happen), `shouldReset.current` is still `false`, so no `RESET` is dispatched. The form stays as-is.
```

## State and authentication

> Activity preserves local component state (`useState`, DOM input values) across navigations, including authentication changes. This is standard React behavior - props changing (such as receiving a new user) triggers a re-render but does not reset existing state. A draft composed by one user shouldn't be visible to another.

Reviewer: ```suggestion
Activity preserves local component state (`useState`, DOM input values) across navigations, including authentication changes. This is standard React behavior: props changing (such as receiving a new user) triggers a re-render but does not reset existing state. A draft composed by one user shouldn't be visible to another.
```

```

Alternatively, key components by user ID to let React handle the reset: `<Form key={userId} />`. See [Resetting state](/docs/app/guides/activity#resetting-state) for more patterns.

## Global styles

> Page-level styles (CSS variables, z-index, global classes) can affect visible pages when the originating component is hidden by Activity. You likely want to disable them when hidden - a hidden page's accent color or z-index overrides shouldn't leak into the visible page.

Reviewer: ```suggestion
Page-level styles (CSS variables, z-index, global classes) can affect visible pages when the originating component is hidden by Activity. You likely want to disable them when hidden: a hidden page's accent color or z-index overrides shouldn't leak into the visible page.
```
